tools/ai_sidecar/server.py

```
@@ -97,6 +119,88 @@ async def _safe_send(websocket: Any, payload: dict[str, Any]) -> None:
+
+    Returns ``None`` when no token is configured (default loopback deployment).
+    Otherwise returns a callable that closes the upgrade with HTTP 401 if the
+    ``X-AC-Copilot-Token`` header is missing or wrong.
+    """
+    if not token:
+        return None
+
+    from websockets.datastructures import Headers
+    from websockets.http11 import Response
+
+    def _check(connection: Any, request: Any) -> Any:
```

Ruff (0.15.10)

[warning] Dynamically typed expressions (typing.Any) are disallowed in `connection`

(ANN401)